File: rtc-sentinel/ml-service/app/quality.py
from typing import Annotated, Literal

from pydantic import BaseModel, Field

QualityLabel = Literal["excellent", "good", "fair", "poor", "critical"]


class QualityFeatures(BaseModel):
    rtt: Annotated[float, Field(strict=True, ge=0)]
    jitter: Annotated[float, Field(strict=True, ge=0)]
    packetLoss: Annotated[float, Field(strict=True, ge=0, le=100)]
    bitrate: Annotated[float, Field(strict=True, ge=0)]
    audioLevel: Annotated[float, Field(strict=True, ge=0, le=1)] = 0.5


class QualityPrediction(BaseModel):
    quality: QualityLabel
    confidence: float = Field(ge=0, le=1)


def _upper_bound_tier(value: float, bounds: tuple[float, ...]) -> int:
    return next((index for index, bound in enumerate(bounds) if value < bound), 4)


def _bitrate_tier(bitrate: float) -> int:
    if bitrate >= 32_000:
        return 0
    if bitrate >= 24_000:
        return 1
    if bitrate >= 16_000:
        return 2
    if bitrate >= 8_000:
        return 3
    return 4


def rule_based_quality_values(
    rtt: float,
    jitter: float,
    packet_loss: float,
    bitrate: float,
) -> QualityLabel:
    tiers = (
        _upper_bound_tier(rtt, (150, 250, 400, 800)),
        _upper_bound_tier(jitter, (20, 40, 70, 120)),
        _upper_bound_tier(packet_loss, (1, 2.5, 5, 10)),
        _bitrate_tier(bitrate),
    )
    labels: tuple[QualityLabel, ...] = ("excellent", "good", "fair", "poor", "critical")
    return labels[max(tiers)]


def predict_rule_quality(features: QualityFeatures) -> QualityPrediction:
    quality = rule_based_quality_values(
        features.rtt,
        features.jitter,
        features.packetLoss,
        features.bitrate,
    )
    return QualityPrediction(quality=quality, confidence=1.0)

File: rtc-sentinel/client/src/quality.ts
import type { QosMetric } from './qos';

export type QualityLabel =
  'Excellent' | 'Good' | 'Fair' | 'Poor' | 'Critical' | 'Unknown';

export interface QualityAssessment {
  quality: QualityLabel;
  score: number | null;
  limitingFactors: string[];
}

const labels: QualityLabel[] = [
  'Excellent',
  'Good',
  'Fair',
  'Poor',
  'Critical',
];
const scores = [100, 80, 60, 40, 20];

function upperBoundTier(value: number, bounds: number[]): number {
  const tier = bounds.findIndex((bound) => value < bound);
  return tier === -1 ? 4 : tier;
}

function bitrateTier(bitrate: number): number {
  if (bitrate >= 32_000) return 0;
  if (bitrate >= 24_000) return 1;
  if (bitrate >= 16_000) return 2;
  if (bitrate >= 8_000) return 3;
  return 4;
}

export function assessCallQuality(
  metric: Pick<QosMetric, 'rtt' | 'jitter' | 'packetLoss' | 'bitrate'>,
): QualityAssessment {
  if (metric.rtt === null || metric.jitter === null) {
    return {
      quality: 'Unknown',
      score: null,
      limitingFactors: ['Waiting for complete WebRTC statistics'],
    };
  }
  const dimensions = [
    {
      name: 'round-trip time',
      tier: upperBoundTier(metric.rtt, [150, 250, 400, 800]),
    },
    { name: 'jitter', tier: upperBoundTier(metric.jitter, [20, 40, 70, 120]) },
    {
      name: 'packet loss',
      tier: upperBoundTier(metric.packetLoss, [1, 2.5, 5, 10]),
    },
    { name: 'bitrate', tier: bitrateTier(metric.bitrate) },
  ];
  const worstTier = Math.max(...dimensions.map((dimension) => dimension.tier));
  return {
    quality: labels[worstTier],
    score: scores[worstTier],
    limitingFactors: dimensions
      .filter((dimension) => dimension.tier === worstTier)
      .map((dimension) => dimension.name),
  };
}
